### PR TITLE
HIVE-25276: Enable automatic statistics generation for Iceberg tables

### PR DESCRIPTION
**Pull request by pvary**
### What changes were proposed in this pull request?
Allow column stat generation when we can commit in a move task

### Why are the changes needed?
So we have column statistics after inserting to an Iceberg table

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Added unit tests

Also added #2359 as it is required for this patch

### CODE REVIEW

**Review by szlta — APPROVED** [review]
+1

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java [hunk 1]
```diff
@@ -1314,6 +1314,85 @@ public void testScanTableCaseInsensitive() throws IOException {
...
+
+    checkColStat(identifier.name(), "customer_id");
+  }
+
+  @Test
+  public void testStatWithInsertOverwrite() {
+    TableIdentifier identifier = TableIdentifier.of("default", "customers");
+
+    shell.setHiveSessionValue(HiveConf.ConfVars.HIVESTATSAUTOGATHER.varname, true);
+    testTables.createTable(shell, identifier.name(), HiveIcebergStorageHandlerTestUtils.CUSTOMER_SCHEMA,
```
Comment (marton-bod) [inline iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:1341]: If we have test cases for unpartitioned insert, partitioned insert, unpartitioned IOW, should we have a test case for partitioned IOW as well?

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java [hunk 2]
```diff
@@ -1314,6 +1314,85 @@ public void testScanTableCaseInsensitive() throws IOException {
...
+
+    if (testTableType != TestTables.TestTableType.HIVE_CATALOG) {
+      // If the location is set and we have to gather stats, then we have to update the table stats now
+      shell.executeStatement("ANALYZE TABLE " + identifier + " COMPUTE STATISTICS FOR COLUMNS");
+    }
+
+    String insert = testTables.getInsertQuery(HiveIcebergStorageHandlerTestUtils.CUSTOMER_RECORDS, identifier, false);
+    shell.executeStatement(insert);
+
+    checkColStat(identifier.name(), "customer_id");
```
Comment (marton-bod) [inline iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:1333]: Are the columns stats gathered for `first_name` and `last_name` as well, we're just saving on the number of describe calls?

Comment (pvary) [reply iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:1333]: They are basically the same, and generated as well.
Only in partitioned case did I went for checking the partition columns too. Just to be sure

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java (hunk 1 above)
Comment (pvary) [reply iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:1341]: My thought process was the following:
- We need a basic test - so created unpartitioned insert
- We need to test partitioned tables (what happens with the partition columns) - so created partitioned test (not sure that this is strictly needed)
- We need to test IOW - so created IOW

I am open to discussions either way

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java (hunk 2 above)
Comment (marton-bod) [reply iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:1333]: Ok, thanks, just wanted to doube-check my understanding.

File: ql/src/java/org/apache/hadoop/hive/ql/parse/ColumnStatsAutoGatherContext.java [hunk 3]
```diff
@@ -71,13 +71,21 @@
...
     this.sa = sa;
     this.conf = conf;
     this.op = op;
-    this.tbl = tbl;
+    try {
+      this.tbl = origTbl.copy();
+    } catch (HiveException he) {
+      throw new SemanticException(he);
+    }
+    if (tbl.getStorageHandler().alwaysUnpartitioned()) {
```
Comment (marton-bod) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/ColumnStatsAutoGatherContext.java:85]: Can the storage handler be null?

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java (hunk 1 above)
Comment (pvary) [reply iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:1341]: Added Partitioned CTAS

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java (hunk 2 above)
Comment (pvary) [reply iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:1333]: Added to a partitioned CTAS test

File: ql/src/java/org/apache/hadoop/hive/ql/parse/ColumnStatsAutoGatherContext.java (hunk 3 above)
Comment (pvary) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/ColumnStatsAutoGatherContext.java]: Not sure, so I have added the `null` check

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/create/CreateTableDesc.java [hunk 4]
```diff
@@ -804,9 +801,18 @@ public Table toTable(HiveConf conf) throws HiveException {
       }
     }
 
-    if (getCols() != null) {
-      tbl.setFields(getCols());
+    Optional<List<FieldSchema>> cols = Optional.of(getCols());
```
Comment (marton-bod) [inline ql/src/java/org/apache/hadoop/hive/ql/ddl/table/create/CreateTableDesc.java:804]: It seems like `getCols()` can return null based on the previous null check. This would fail here as well, unless wrap it using `Optional.ofNullable()`

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java [hunk 5]
```diff
@@ -1314,6 +1314,104 @@ public void testScanTableCaseInsensitive() throws IOException {
...
+  @Test
+  public void testStatWithPartitionedCTAS() {
+    Assume.assumeTrue(HiveIcebergSerDe.CTAS_EXCEPTION_MSG, testTableType == TestTables.TestTableType.HIVE_CATALOG);
+
+    shell.executeStatement("CREATE TABLE source (id bigint, name string) PARTITIONED BY (dept string) STORED AS ORC");
+    shell.executeStatement(testTables.getInsertQuery(
+        HiveIcebergStorageHandlerTestUtils.CUSTOMER_RECORDS, TableIdentifier.of("default", "source"), false));
+
+    shell.setHiveSessionValue(HiveConf.ConfVars.HIVESTATSAUTOGATHER.varname, true);
+
```
Comment (marton-bod) [inline iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:1398]: nit: extra new line

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/create/CreateTableDesc.java [hunk 6]
```diff
@@ -804,9 +801,18 @@ public Table toTable(HiveConf conf) throws HiveException {
...
+    Optional<List<FieldSchema>> partCols = Optional.of(getPartCols());
+
+    if (storageHandler !=null && storageHandler.alwaysUnpartitioned()) {
```
Comment (marton-bod) [inline ql/src/java/org/apache/hadoop/hive/ql/ddl/table/create/CreateTableDesc.java:807]: nit: `!= null`

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/create/CreateTableDesc.java (hunk 4 above)
Comment (pvary) [reply ql/src/java/org/apache/hadoop/hive/ql/ddl/table/create/CreateTableDesc.java]: Thanks!
Done

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/create/CreateTableDesc.java (hunk 6 above)
Comment (pvary) [reply ql/src/java/org/apache/hadoop/hive/ql/ddl/table/create/CreateTableDesc.java]: Done

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java (hunk 5 above)
Comment (pvary) [reply iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java]: Done

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java [hunk 7]
```diff
@@ -233,15 +237,21 @@ public void preAlterTable(org.apache.hadoop.hive.metastore.api.Table hmsTable, E
...
+      if (hmsTable.isSetPartitionKeys() && !hmsTable.getPartitionKeys().isEmpty()) {
+        List<PartitionTransformSpec> spec = PartitionTransform.getPartitionTransformSpec(hmsTable.getPartitionKeys());
+        if (!SessionStateUtil.addResource(conf, hive_metastoreConstants.PARTITION_TRANSFORM_SPEC, spec)) {
+          throw new MetaException("Query state attached to Session state must be not null. " +
+              "Partition transform metadata cannot be saved.");
+        }
         hmsTable.getSd().getCols().addAll(hmsTable.getPartitionKeys());
         hmsTable.setPartitionKeysIsSet(false);
       }
+      preAlterTableProperties.spec = spec(conf, preAlterTableProperties.schema, hmsTable);
```
Comment (marton-bod) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:253]: Is this needed here? Or if we use it only for the validation inside `spec()`, then maybe remove the local variable, or better yet, extract the validation logic into a different method we can call here?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java [hunk 8]
```diff
@@ -233,15 +237,21 @@ public void preAlterTable(org.apache.hadoop.hive.metastore.api.Table hmsTable, E
...
       preAlterTableProperties.schema = schema(catalogProperties, hmsTable);
-      preAlterTableProperties.spec = spec(conf, preAlterTableProperties.schema, catalogProperties, hmsTable);
       preAlterTableProperties.partitionKeys = hmsTable.getPartitionKeys();
 
       context.getProperties().put(HiveMetaHook.ALLOW_PARTITION_KEY_CHANGE, "true");
       // If there are partition keys specified remove them from the HMS table and add them to the column list
-      if (hmsTable.isSetPartitionKeys()) {
+      if (hmsTable.isSetPartitionKeys() && !hmsTable.getPartitionKeys().isEmpty()) {
+        List<PartitionTransformSpec> spec = PartitionTransform.getPartitionTransformSpec(hmsTable.getPartitionKeys());
+        if (!SessionStateUtil.addResource(conf, hive_metastoreConstants.PARTITION_TRANSFORM_SPEC, spec)) {
```
Comment (marton-bod) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:246]: Where was this part before that we saved it into the session conf?

File: iceberg/iceberg-handler/src/test/results/positive/vectorized_iceberg_read.q.out [hunk 9]
```diff
@@ -129,17 +129,17 @@ Stage-0
     Stage-1
       Reducer 2 vectorized
       File Output Operator [FS_11]
-        Select Operator [SEL_10] (rows=1 width=564)
+        Select Operator [SEL_10] (rows=1 width=372)
```
Comment (marton-bod) [inline iceberg/iceberg-handler/src/test/results/positive/vectorized_iceberg_read.q.out:132]: Out of curiosity: do you know why the width has descreased so much?

File: ql/src/test/org/apache/hadoop/hive/ql/exec/tez/TestTezOutputCommitter.java [hunk 10]
```diff
@@ -122,6 +122,7 @@ private IDriver getDriverWithCommitter(String committerClass) {
     conf.setVar(HiveConf.ConfVars.HIVE_AUTHORIZATION_MANAGER,
         "org.apache.hadoop.hive.ql.security.authorization.plugin.sqlstd.SQLStdHiveAuthorizerFactory");
     conf.setBoolVar(HiveConf.ConfVars.HIVE_SUPPORT_CONCURRENCY, false);
+    conf.setBoolVar(HiveConf.ConfVars.HIVESTATSCOLAUTOGATHER, false);
```
Comment (szlta) [inline ql/src/test/org/apache/hadoop/hive/ql/exec/tez/TestTezOutputCommitter.java:125]: Why is this required here?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java (hunk 7 above)
Comment (pvary) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:253]: This is moved from line 236. We need it to be set, but we have to do it after we got the correct spec

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java (hunk 8 above)
Comment (pvary) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:246]: This is for migrating tables from non-Iceberg tables to Iceberg tables. Previously we just depended on the partition cols, from now on we need to have the data in the `SessionState` instead. So we put that there

File: iceberg/iceberg-handler/src/test/results/positive/vectorized_iceberg_read.q.out (hunk 9 above)
Comment (pvary) [reply iceberg/iceberg-handler/src/test/results/positive/vectorized_iceberg_read.q.out:132]: TBH I am not sure, but I expect that has something to do with the new statistics

File: ql/src/test/org/apache/hadoop/hive/ql/exec/tez/TestTezOutputCommitter.java (hunk 10 above)
Comment (pvary) [reply ql/src/test/org/apache/hadoop/hive/ql/exec/tez/TestTezOutputCommitter.java:125]: Otherwise the tests are failing, because with stats turned on we generate 2 tasks instead of 1 (change of the execution plans which contain a stage)

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java (hunk 7 above)
Comment (marton-bod) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:253]: Right, thanks

**Review by marton-bod — APPROVED** [review]
